packages/modules/openwb_flex/device.py

```diff
@@ -28,25 +30,49 @@ def get_default_config() -> dict:
     }
 
 
-class Device(abstract_device.AbstractDevice):
+class Device(AbstractDevice):
     COMPONENT_TYPE_TO_CLASS = {
         # "bat": ,
         "counter": counter.EvuKitFlex,
         "inverter": inverter.PvKitFlex
     }
+    _components = []  # type: List[counter.EvuKitFlex]
 
-    def __init__(self, device: dict) -> None:
+    def __init__(self, device_config: dict) -> None:
         try:
-            ip_address = device["configuration"]["ip_address"]
-            port = device["configuration"]["port"]
-            client = modbus.ModbusClient(ip_address, port)
-            super().__init__(device, client)
+            self.device_config = device_config
+            ip_address = device_config["configuration"]["ip_address"]
+            port = device_config["configuration"]["port"]
+            self.client = modbus.ModbusClient(ip_address, port)
         except Exception:
-            log.MainLogger().exception("Fehler im Modul "+device["name"])
+            log.MainLogger().exception("Fehler im Modul " + device_config["name"])
 
     def add_component(self, component_config: dict) -> None:
-        self.instantiate_component(component_config, super(
-        ).component_factory(component_config["type"]))
+        component_type = component_config["type"]
+        if component_type in self.COMPONENT_TYPE_TO_CLASS:
+            self._components.append(self.COMPONENT_TYPE_TO_CLASS[component_type](
+                self.device_config["id"], component_config, self.client))
+
+    def get_values(self) -> None:
+        log.MainLogger().debug("Start device reading" + str(self._components))
+        if self._components:
+            for component in self._components:
+                try:
+                    component.get_values()
+                except ModuleError as e:
+                    e.store_error(component.get_component_info())
+                except Exception as e:
+                    ModuleError(
+                        str(type(e)) + " " + str(e),
+                        ModuleErrorLevel.ERROR).store_error(
+                        component.get_component_info())
+                else:
+                    ModuleError("Kein Fehler.", ModuleErrorLevel.NO_ERROR).store_error(component.get_component_info())
+        else:
+            log.MainLogger().warning(
+                self.device_config["name"] +
+                ": Es konnten keine Werte gelesen werden, da noch keine Komponenten konfiguriert wurden."
+            )
```

Comment:
Ich finde der Klasse `Device`, die es dann in Zukunft hundertfach geben wird, wird hier zu viel zugemutet.

Ich finde besonders diese Zeile: `ModuleError("Kein Fehler.", ModuleErrorLevel.NO_ERROR).store_error(component.get_component_info())` nicht so intuitiv, weil es sehr wiedersprüchlich ist worum es geht. Also `ModuleError("Kein Fehler.")` ist merkwürdig.

Ich frage mich aber auch bei den anderen Zeilen, ob diese nicht ersatzlos gestrichen werden können. Die Exception enthält einen kompletten Stacktrace und damit kann der verursachende Code gefunden werden.

Aber gut, wenn ich das so sehe soll auch ein Status auf der UI landen oder sowas. Dann wäre zumindest die Component-ID sinnvoll. Dafür reicht dann aber auch:
```python
try:
    do_stuff()
except Exception e:
    raise ComponentUpdateError(self, component) from e
```
Dann hat der Exceptionhandler alles was er braucht.

Reply:
Das Exception-Handling der Komponente gehört eigentlich nicht auf Device-Ebene. Da es allerdings keine allgemeine Komponenten-Klassen gibt, war die Idee, die Exceptions beim Auslesen der jeweiligen Komponenten auf Device-Ebene abzufangen.
Statt `ModuleError` wäre `FaultState` vielleicht treffender. 

Das ganze soll etwa so ablaufen: 
Ein ModuleError wird geworfen, wenn der Fehler bekannt ist, z.B. ein ConnectionError lässt darauf schließen, dass die IP-Adresse falsch ist oder das Gerät falsch angeschlossen ist. Dann wird ein ModuleError mit einer Endbenutzerfreundlichen Fehlermeldung geworfen. Die Fehlermeldung wird auf der Status-Seite bei der jeweiligen Komponente angezeigt und noch einmal mit Stacktrace ins Log geschrieben. Wenn an anderer Stelle eine Exception geworfen wird, soll diese ebenfalls als Fehler dem Endbenutzer angezeigt werden (mit dem Text der Python-Meldung), zB. für Forumssuche. Wenn kein Fehler aufgetreten ist, muss der Fehlerstatus auf dem Broker auf kein Fehler gesetzt werden.
Wenn bei einer Komponente ein Fehler aufgetreten ist, sollen trotzdem allen anderen Komponenten des Devices ausgelesen werden.

Reply:
# Fehlende automatische Auswertbarkeit der Fehlermeldung
Ein `raise Exception("bla bla bla, hier ist mein Text")` mag im Log erstmal helfen, aber es ist sehr schwer die Exception selbst zu analysieren, denn sie hat keine Struktur.

Naheliegendstes Beispiel dafür wäre Internationalisierung. Wenn irgendwann die UI von oWB doch mal mehrsprachig sein soll, dann verbaut man sich den Weg dahin, dass die Fehlermeldungen es auch sind, denn die sind leider in Textform hier und das kann man kaum rückgängig machen.

Besser ist es konkrete Exceptions zu werfen, die sich selbst serialisieren können. Zum Beispiel so:
```python
class SerializableException:
    @abstractmethod
    def to_dict(self) -> dict:
        pass


class ConnectionError(Exception, SerializableException):
    def __init__(self, host: str, port: int):
        self.host = host
        self.port = port
        
    def to_dict(self):
        return {"type": "CONNECTION_ERROR", "host": self.host, "port": self.port}
```
Dann kann man, wenn man den Fehler an die UI weitergeben will die Methode `to_dict` aufrufen, das Ergebnis zu einem JSON serialisieren und dann kann sich die UI überlegen, was sie damit tut. Zum Beispiel das ganze als Text darstellen, aber in unterschiedlicher Sprache je nach Einstellung:
```javascript
function print_error(error) {
    if (error.type == 'CONNECTION_ERROR') {
        if (language == 'de') {
            return `Verbindung zu ${error.host}:${error.port} fehlgeschlagen. Probiere nochmal in der nächsten Mondphase.`
        }
        if (language == 'en') {
            return `Connection to ${error.host}:${error.port} failed. Try again during the next phase of the moon.`
        }
    }
}
```
(In der Praxis das ganze natürlich nicht in ehlendig riesigen `if`-statements, sondern aus einer passenden Übersetzungsdatei eine formatierbare Vorlage ziehen, aber vom Prinzip eben so.

Dieses strukturierte Vorgehen macht es auch viel einfacher andere Dinge zu ermöglichen. Es ist zum Beispiel nicht notwendigerweise nur Text im String. Auf der UI der openWB könnten auch Bilder sein oder Links zur Hilfe oder Links zur passenden Konfigurationsseite. Es ist aber auch möglich, dass der Fehler strukturiert geloggt wird, also zum Beispiel in einer Datenbank und dann kann man statistisch auswerten, wie häufig der raus kommt. Und natürlich greift nicht nur die offizielle UI von oWB auf die Fehler zu, sondern auch andere Leute mit ihren Automatisierungsskripten per MQTT. Mit so einer strukturierten Fehlermeldung kann man gleich richtig viel mehr anfangen.

# Fehlerstatus resetten
> Wenn kein Fehler aufgetreten ist, muss der Fehlerstatus auf dem Broker auf kein Fehler gesetzt werden.

Ja, aber das muss nicht einzeln sein. Das geht auch automatisch für alle zusammen. Es gibt irgendwo Code, der die Methode `get_values` aufruft. Der Code kann das übernehmen:
```python
try:
    device.get_values()
    clear_all_error_states(device)
except Exception e:
    set_error_state(e)
```

# Wenn eine Komponente noch geht...
> Wenn bei einer Komponente ein Fehler aufgetreten ist, sollen trotzdem allen anderen Komponenten des Devices ausgelesen werden.

Warum? Manche Komponenten haben sogar Verrechnungen. Wenn zwei Zähler für zwei Wechselrichter da sind und der eine Wechselrichter weg ist und man die Werte nicht addieren kann, was dann? Einfach nur den einen Wert nehmen ist doof, das zerschießt die ganze Statistik. Selbst wenn die Werte nicht in Bezug zueinander stehen: Wenn etwas falsch konfiguriert ist, ist es wirklich hilfreich die Informationen sind nur teilweise angezeigt? Theoretisch ist das natürlich denkbar, aber wie wahrscheinlich ist das? Letztlich muss der Nutzer sowieso seine Konfig fixen. Wieviel Aufwand will man betreiben um aus kaputten Daten noch ein bißchen was rauzuholen? Es erhöht auch massiv die Testfälle die abgedeckt werden müssen. Also man muss dann auch für jedes einzelne Modul mal ausprobieren was passiert, wenn nur ein Teil der Daten da ist. Oder man risikiert, dass die teilweisen Daten ohnehin nichts nützen. Ich würde behaupten, dass ist der Mehraufwand nicht wert.

Wenn es aber unbedingt so gewollt ist, dann das ganze in eine Hilfsfunktion auslagern:
```python
def update_component(component, update_function: Callable):
    try:
        update_function()
    except ModuleError as e:
        e.store_error(component.get_component_info())
    except Exception as e:
        # ...

for component in self._components:
    update_component(component, lambda: component.get_values())
```
Die Funktion `update_component` könnte dann irgendwo ins common-Paket oder so.

Reply:
Das Error Handling für die Komponenten auszulagern ist sinnvoll.

Ich finde, wenn die Werte nicht miteinander verrechnet werden, ist es trotzdem sinnvoll die Werte der funktionierenden Komponenten abzufragen, selbst wenn bei einer ein Fehler auftritt. 
Die WR-Werte haben auf die Regelung ohnehin keinen Einfluss. Warum also dann nicht weiterhin mit den Werten des EVU-Zählers regeln, auch wenn der WR nicht erreichbar ist? Bei Modulen, wo die Werte miteinander verrechnet werden, macht das natürlich keinen Sinn. Das kann individuell in der `get_values`-Methode des Devices abgebildet werden. 
Sollen bei einem Fehler in einer Komponente trotzdem die anderen abgefragt werden:
```Python
for component in self._components:
    try:
        component.update()
    except Exception as e:
        process_component_error(e, component)
```
Hängen die Komponenten zusammen und im Fehlerfall können gar keine Werte ausgegeben werden:
```Python
try:
    for component in self._components:
        component.update()
except Exception as e:
    for component in self._components:
        process_component_error(e, component)
```
Die Fehlermeldung muss dann für alle Komponenten ausgegeben werden, damit für den Benutzer ersichtlich ist, warum die Werte nicht aktualisiert werden.

Reply:
Ich habe den Eindruck, dass du meine Argumentation nicht ganz verstanden hast. Deswegen nochmal in aller Kürze: Ja, du hast Recht, alle Komponenten zu testen ist schöner. Ich wollte anregen, dass Feature dennoch wegzulassen, weil es Arbeit verursacht, und ich glaube dass es nur sehr selten in wenigen Nischenfällen hilfreich ist. Andere Dinge erscheinen mir wichtiger.

Aber gut, für alles Weitere nehme ich an, dass die Entscheidung es mit "alle Komponenten probieren" eine gefällte Entscheidung ist und es so gemacht werden soll.

Ich habe noch etwas darüber nachgedacht, wie man den Fehlerfall möglichst einfach machen kann. Ich schätze, dass viele Modulentwickler keine Python-Profis (oder überhaupt Programmierprofis) sind und die Gefahr besteht, dass diese das Konstrukt nicht verstehen. Das führt dann dazu, dass der Code von Modul zu Modul kopiert wird und eventuell passt der Code dann aber garnicht mehr. Also wie kann man es möglichst einfach machen. Eine Idee ist mir gerade noch gekommen: Einen Context-Manager verwenden:
```python
class ComponentUpdateContext:
    def __init__(self, device, component):
        self.device = device
        self.component = component

    def __enter__(self):
        return None

    def __exit__(self, exception_type, exception, exception_traceback) -> bool:
        if exception is None:
            reset_component_error_state(self.device, self.component)
            return True
        set_component_error(self.device, self.component, exception)
        return True
```
Zum Vergleich:
```python
# OHNE context manager:
# ----------------------------
for component in self._components:
    try:
        component.update()
    except Exception as e:
        process_component_error(e, component)


# Mit context manager:
# ----------------------------
for component in self._components:
    with ComponentUpdateContext(self, component):
        component.update()
```
Das ist etwas ähnlich zu meinem Vorschlag mit der Funktion `update_component` weiter oben, glaube aber das System so dürfte noch etwas verständlicher sein.